paf-mvp-operator-express/tests/integration-tests/operator.integration.test.ts

```diff
@@ -1,4 +1,10 @@
-import { assertError } from '../helpers/integration.helpers';
+import {
```

Comment:
I am not really comfortable with the changes within this file: It adds currying functions and "if condition" in tests for the redirect. It seems more complex than it should.

Tests should be deceptively simple and straightforward to read. Can we simplify those ones?

Reply:
I don't consider currying functions as complex, I personally think they are quite readable, but don't hesitate to propose a simpler option (the only option I think about is adding an `if` statement, which is your second concern about complexity! 😅).
Regarding `if` statements, I agree with you as I'm not 100% happy to add conditions in tests, however... a **lot** of the test is similar between the rest and the redirect tests, and I even believe it's important it is visible in the tests themselves.
So, the only other option I see is to duplicate the whole test suite, and that would be a big regression. Do you have something else in mind?

Reply:
Please note the nuance between complex and _more complex than it should._ There is no additional value of the currying function here because we call directly the function each time. A simple extra-parameter in the function would do the job.

Concerning code duplication, there is a school of thought that laxes the duplication track in test code: having malleable and readable tests has a higher priority than code factorization. I am comfortable with a bit of duplication if it brings clarity and I don't consider it a regression. 

Here are a set of suggestions for improving readability:
1. Just add simple parameters instead of currying functions.
2. Separate the verification of the `startMock`and `endMock` from the rest of the tests. It creates noise in the tests. It will create a bit of duplication, yes. But the readability of the tests that check the responses will drastically increase as you only check one thing each time. Better separation of concern IMHO.
3. Improve our custom assert of `Response` by including the verification of returned URL in case of redirection. It will hide the `if` in the test and it is proper to create custom assert that tests a whole object.
4. Split `should handle valid request` so that we don't test one call and then another. Separation of concern.
5. Consider a custom assert for `GetIdsPrefsResponse` or the creation of the full object and check the equality.
6. Put the private key of `'for wrong signature'` as a constant at the top of the file or test suite for making the test readable. Or consider improving the `ClientBuilder`.
7. I didn't dig if it is doable here (Are generated signatures deterministic output?) but consider creating expected request and checking the equality instead of checking piece by piece.
8. Create a function with an expressive name that wraps `supertest(server).get(url).set('referer', refererUrl).set('Origin', refererUrl)`

Reply:
Thanks.

New commit pushed with fixes for:

- [x] 1. done, thanks
- [ ] 2. Unless I don't understand your suggestion, it would not add "a bit of duplication", but make _all_ test cases duplicated: we want to test that both handlers are called in **all** cases: it's even a bug we need to fix already today, as some cases don't trigger the end span. If you have something else in mind, please give details.
- [x] 3. I preferred to leave it in a second method to keep the assertError with a single responsibility, which is the same for both rest and redirect. But I got rid of the `if`.
- [ ] 4. I'm not sure to follow you. Do you mean split to have one for rest and one for redirect? In this case, I only see two options: either to duplicate the test (with worse maintainability), or create sub-functions that do the common assertions, with more indirection. Am I missing a third way?
- [x] 5. Added a test for a known user with cookies, good point. However these are integration tests, I left the assertion to a simple check of a few values, not deep equality.
- [x] 6. Added a constant (although I don't find it much more readable. Not sure this is what you had in mind)
- [ ] 7. See 5: I don't think this is the place to test for deep equality.
- [ ] 8. As you pointed out yourself, in tests I sometimes prefer duplication for better readability. In this case I think the grammar of supertest (with method chaining) is pretty explicit so I find it readable. Notice that there is a lot of variation in the different tests so I don't really see how a function would make it more readable?

Reply:
Thanks @MohamedAmineGhazouani .
Concerning the `verifyRedirectUrl()`, it's new. Please check 3 commits in the past, there was an `if` around all these calls (which is verbose, but more explicit). This corresponds to change number 3 in the list above.
I think I prefer with the `if` statement, because I agree with you, I dislike the call to a method that does nothing 1/2 of the time. Maybe I could rename it to something like `verifyRedirectUrlIfNeeded()`? Not really convincing either :-(

Reply:
The last commits focused only on minor suggestions or tweaking the major ones to make the code worse. Therefore, I don't see many improvements. However, I don't want we spend too much time on this thread. So I will clarify some parts of the discussion here but I am validating the PR. We can discuss this offline for simplicity. It will be your call.

Concerning, the 2/, first it can be factorized with cases constant including lambdas. Then, we don't need to test all the cases for the spans. As a rule of thumb, integration tests are in the middle of the test pyramid and their purpose is to verify that the different modules are well integrated with each other. If we feel the need to test everything in integration tests then it is a code smell.

Concerning your last messages and the 3/, `verifyRedirectUrl()` is confusing and is worst than the `if`.  If you aren't interested in a custom assertion that tests the whole response (I don't see any issue about separation of concern), then dish this last commit.

Concerning the 4/, I think that the code confused me because I saw initially a pattern "Arrange, Act, Assert, Act, Assert" instead of the classic "Arrange, Act, Assert (AAA)". The point is that the 'Assert' part is too big and handles too many cases with the use of intermediary functions. This is the root of the confusion. So maybe it requires a split or a simplification. You can reuse the existing suggestions (e.g custom asserts, splitting, etc).